simpl/incubator/rest.py

```diff
@@ -190,3 +154,41 @@ def wrapped(*args, **kwargs):
                 bottle.abort(400, str(exc))
         return wrapped
     return deco
+
+
+def body(schema=None, types=None, required=False, default=None):
+    """Decorator to parse and validate API body.
+
+    :keyword schema: callable that accepts raw data and returns the coerced (or
+        unchanged) data if it is valid. It should raise an error if the data is
+        not valid.
+    :keyword types: supported content types (default is ['application/json'])
+    :keyword required: if true and no body specified will raise an error.
+    :keyword default: default value to return if no body supplied
+
+    Note: only json types are supported.
+    """
+    if not types:
+        types = ['application/json']
+    if not all('json' in t for t in types):
+        raise NotImplementedError("Only 'json' body supported.")
+
+    def wrap(fxn):
+        """Return a decorated callable."""
+        def wrapped(*args, **kwargs):
+            """Callable to called when the decorated function is called."""
+            body = bottle.request.json
+            if required and not body:
+                bottle.abort(400, "Call body cannot be empty")
+            if body is None:
+                body = default
+            if schema:
+                try:
+                    body = schema(body)
+                except (KeyboardInterrupt, SystemExit):
+                    raise  # don't catch and ignore attempts to end the app
+                except Exception as exc:
+                    bottle.abort(400, str(exc))
+            return fxn(*args, body=body, **kwargs)
```

Comment:
It seems like some amount of this discussion is about signatures.

We expect to need to use bottle's threadlocal-based interface for accessing at least _some_ of the current request context, no matter what it attribute/value in that context we are going after. As much as I'd like for things to be explicit, this is how bottle is used.

Could we just allow the decorators to do the validation as well as mutating those globals? Then we go about our business accessing whichever pieces of the request context as usual, from these:

https://github.com/bottlepy/bottle/blob/0.12.8/bottle.py#L3691-L3698

I think this interface would be more consistent with "how to use bottle", and should relieve some of the function signature problems?